### PR TITLE
[Feature] Bitbucket Addon [OSF-8192]

### PR DESCRIPTION
**Pull request by mfraezz**
Updates #6954 

> ## Purpose
> 
> Add support for a read-only bitbucket OSF add-on.  Still targets `develop-backup` and needs to be upgraded to work with DjangOSF.
> 
> ## Changes
> 
> Adds a whole new provider in `website/addons/bitbucket`.  All changes outside that directory go directly towards supporting the new addon.
> 
> ## Side effects
> 
> Unavoidable grief.
> 
> ## Ticket
> 
> [#SVCS-268] - [JIRA](https://openscience.atlassian.net/browse/SVCS-268)
> [#OSF-8192] - [JIRA](https://openscience.atlassian.net/browse/OSF-8192)
> 
> Entangled with WB PR: https://github.com/CenterForOpenScience/waterbutler/pull/198

# MIGRATION WARNING
Not only does this share a migration number (`0042`) with #7418 , that migration has a conflicting action -- altering possible types for `BaseFileNode`. 
### **Running `makemigrations --merge` will NOT be sufficient.** 
Instead, whichever PR goes in last will need the migration recreated before being merged.

### CODE REVIEW

File: addons/bitbucket/views.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    changed = (
+        bitbucket_user_name != node_settings.user or
+        bitbucket_repo_name != node_settings.repo
+    )
+
+    # Update hooks
+    if changed:
+
+        # # Delete existing hook, if any
+        # node_settings.delete_hook()
```
Comment (mfraezz) [inline addons/bitbucket/views.py]: Note: Bitbucket does support webhooks, but Services/Product teams opted to skip them in the short term to release this faster. Comments left to simplify adding in future

File: api/base/settings/defaults.py [hunk 2]
```diff
@@ -277,7 +278,7 @@
 ESI_MEDIA_TYPES = osf_settings.ESI_MEDIA_TYPES
 
 ADDONS_FOLDER_CONFIGURABLE = ['box', 'dropbox', 's3', 'googledrive', 'figshare', 'owncloud']
-ADDONS_OAUTH = ADDONS_FOLDER_CONFIGURABLE + ['dataverse', 'github', 'mendeley', 'zotero', 'forward']
+ADDONS_OAUTH = ADDONS_FOLDER_CONFIGURABLE + ['dataverse', 'github', 'bitbucket', 'mendeley', 'zotero', 'forward']
```
Comment (mfraezz) [inline api/base/settings/defaults.py]: API tests for this need to be added

File: framework/addons/data/addons.json [hunk 3]
```diff
@@ -291,6 +291,36 @@
...
+                "status": "none",
+                "text": "Bitbucket does not support adding or updating files via their API."
+            },
+            "Delete files": {
+                "status": "none",
+                "text": "Bitbucket does not support deleting files via their API."
+            },
+            "Logs": {
+                "status": "full",
+                "text": "Bitbucket dynamically updates OSF logs when files are modified outside the OSF. Changes to Bitbucket repos made before the repo is linked to the OSF will not be reflected in OSF logs."
```
Comment (mfraezz) [inline framework/addons/data/addons.json]: Untrue, might be true in future

**Comment by felliott** [comment]
The corresponding WB PR has been merged and pushed to [WB's `develop` branch](https://github.com/CenterForOpenScience/waterbutler/tree/develop).

File: website/static/js/logTextParser.js [hunk 4]
```diff
@@ -495,6 +495,17 @@ var LogPieces = {
         }
     },
 
+    bitbucket_repo: {
+        view: function(ctrl, logObject) {
+            var bitbucket_user = logObject.attributes.params.bitbucket_user;
```
Comment (sloria) [inline website/static/js/logTextParser.js]: Minor: as per the style guidelines, use `lowerCameCase` for variable names.

I'm aware that the github code is inconsistent--that can be changed another day...

**Comment by mfraezz** [comment]
@sloria , finagling complete.

**Pull request merged**